client/components/date-range/index.js

```diff
@@ -0,0 +1,526 @@
+	 * Typically required when the user makes a selection but then dismisses
+	 * the DateRange without clicking "Apply"
+	 */
+	revertDates() {
+		this.setState( previousState => {
+			const newState = { staleDatesSaved: false };
+
+			if ( previousState.staleStartDate && previousState.staleEndDate ) {
+				newState.startDate = previousState.staleStartDate;
+				newState.endDate = previousState.staleEndDate;
+				newState.textInputStartDate = this.formatDateToLocale( this.state.staleStartDate );
+				newState.textInputEndDate = this.formatDateToLocale( this.state.staleEndDate );
+			}
+
+			return newState;
+		} );
+	}
+
+	/**
+	 * Converts a moment date to a native JS Date object
+	 * @param  {MomentJSDate} momentDate a momentjs date object to convert
+	 * @return {DATE}            the converted JS Date object
+	 */
+	momentDateToJsDate( momentDate ) {
+		return this.props.moment.isMoment( momentDate ) ? momentDate.toDate() : momentDate;
+	}
+
+	/**
+	 * Converts a native JS Date object to a MomentJS Date object
+	 * @param  {Date} nativeDate date to be converted
+	 * @return {MomentJSDate}            the converted Date
+	 */
+	nativeDateToMoment( nativeDate ) {
+		return this.props.moment( nativeDate );
+	}
+
+	/**
+	 * Formats a given date to the appropriate format for the
+	 * current locale
+	 * @param  {Date|MomentJSDate} date the date to be converted
+	 * @return {String}      the date as a formatted locale string
+	 */
+	formatDateToLocale( date ) {
+		return this.props.moment( date ).format( 'L' );
+	}
+
+	/**
+	 * 	Gets the locale appropriate date format (eg: "MM/DD/YYYY")
+	 * @return {String} date format as a string
+	 */
+	getLocaleDateFormat() {
+		return this.props.moment.localeData().longDateFormat( 'L' );
+	}
+
+	/**
+	 * Enforces that given date is within the bounds of the
+	 * range specified
+	 * @param  {Moment} date             momentJS instance
+	 * @param  {Moment|Date} options.dateFrom the start of the date range
+	 * @param  {Moment|Date} options.dateTo   the end of the date range
+	 * @return {Moment}                  the date clamped to be within the range
+	 */
+	clampDateToRange( date, { dateFrom, dateTo } ) {
+		// Ensure endDate is within bounds of firstSelectableDate
+		if ( dateFrom && date.isBefore( dateFrom ) ) {
+			date = dateFrom;
+		}
+
+		if ( dateTo && date.isAfter( dateTo ) ) {
+			date = dateTo;
+		}
+
+		return date;
+	}
+
+	/**
+	 * Builds an appropriate disabledDays prop for DatePicker
+	 * based on firstSelectableDate and lastSelectableDate
+	 * config props
+	 *
+	 * See:
+	 * http://react-day-picker.js.org/api/DayPicker/#disabledDays
+	 * http://react-day-picker.js.org/docs/matching-days
+	 *
+	 * @return {array} configuration to be passed to DatePicker as disabledDays prop
+	 */
+	getDisabledDaysConfig() {
+		const { firstSelectableDate, lastSelectableDate } = this.props;
+
+		let config = {};
+
+		if ( firstSelectableDate ) {
+			config = {
```

Comment:
Rather than reassigning the value of `config` we could use regular property assignment here:

```js
const config = {};

if ( firstSelectableDate ) {
  config.before = ...
}
if ( lastSelectableDate ) {
  config.after = ...
}
```

We could even go sort of [point-free](https://wiki.haskell.org/Pointfree) with:

```js
return [ {
  ...firstSelectableDate && { before: .... }, 
  ...lastSelectableDate && { after: .... },
} ]
```

But I still think the 2 `if`s read better

Reply:
I must have been tired when writing this code! I'll go with the simple `if/else` as it's much clearer.

Reply:
I like that you were thinking of immutability though ;)